// @sendbird/uikit-react/Thread/components/ThreadList.js
import { _ as __assign } from '../../chunks/bundle-DseT8u_-.js';
import React__default from 'react';
import ThreadListItem from './ThreadListItem.js';
import { compareMessagesForGrouping } from '../../Channel/utils/compareMessagesForGrouping.js';
import { isSameDay } from 'date-fns';
import { MessageProvider } from '../../Message/context.js';
import { g as getCaseResolvedReplyType } from '../../chunks/bundle---aixLy5.js';
import { u as useThread } from '../../chunks/bundle-Cfzmq-qy.js';
import { u as useSendbird } from '../../chunks/bundle-B_GgIa4W.js';
import { u as useLocalization } from '../../chunks/bundle-eyKmYzrx.js';
import '../../chunks/bundle-DT0pStRO.js';
import '../../chunks/bundle-C__iJnzl.js';
import '../../chunks/bundle-BJzhuBtM.js';
import '../../ui/DateSeparator.js';
import '../../chunks/bundle-CO1y7D0u.js';
import '../../chunks/bundle-fMVQGLVC.js';
import '../../chunks/bundle-ubX26fzL.js';
import '../../chunks/bundle-gXXMGw5_.js';
import '../../chunks/bundle-BtliqdNT.js';
import 'react-dom';
import '../../chunks/bundle-DVYMPV3i.js';
import '../../chunks/bundle-KhnlGrVI.js';
import '../../ui/IconButton.js';
import '../../ui/Button.js';
import '../../ui/Icon.js';
import '../../chunks/bundle-DTKm24fD.js';
import '../../chunks/bundle-B9mN-BTk.js';
import '../../chunks/bundle-DBSWrD1e.js';
import '@sendbird/chat/groupChannel';
import '../../utils/message/getOutgoingMessageState.js';
import '../../chunks/bundle-D1D0Y4xo.js';
import '../../chunks/bundle-DirGjNLX.js';
import '../../chunks/bundle-1qCVlMXW.js';
import '../../ui/FileViewer.js';
import '../../hooks/useModal.js';
import '../../chunks/bundle-CafUCQoC.js';
import '@sendbird/chat';
import '@sendbird/chat/openChannel';
import '../../chunks/bundle-BUm1uq9-.js';
import '../../ui/ImageRenderer.js';
import '../../chunks/bundle-DLAbAE6W.js';
import '../../chunks/bundle-CMjQ-cCa.js';
import '@sendbird/uikit-tools';
import '../../chunks/bundle-B8m-rKye.js';
import '../../chunks/bundle-ClgG8uMn.js';
import '../../chunks/bundle-0TrdjZtT.js';
import '../../chunks/bundle-DhyxF0Xw.js';
import '../../ui/MessageInput.js';
import '../../chunks/bundle-4VxKLKl_.js';
import 'dompurify';
import '../../chunks/bundle-C393fElF.js';
import '../../chunks/bundle-DORSLJqP.js';
import '../../chunks/bundle-CT-NzRzq.js';
import '../../chunks/bundle-AgGVGhWf.js';
import '../context/types.js';
import '../../chunks/bundle-QNHGsSZN.js';
import '../../ui/SortByRow.js';
import '../../ui/MessageItemReactionMenu.js';
import '../../ui/ReactionButton.js';
import '../../chunks/bundle-BWTxQFSl.js';
import '../../chunks/bundle-BUc3HZ08.js';
import '../../ui/Loader.js';
import '../../chunks/bundle-DG9TZERe.js';
import '../../chunks/bundle-DuQ2vnIR.js';
import '../../ui/EmojiReactions.js';
import '../../ui/ReactionBadge.js';
import '../../ui/BottomSheet.js';
import '../../ui/UserListItem.js';
import '../../chunks/bundle-BCWy1eRp.js';
import '../../ui/MutedAvatarOverlay.js';
import '../../ui/Checkbox.js';
import '../../ui/UserProfile.js';
import '../../sendbirdSelectors.js';
import '../../chunks/bundle-LBCBjs9d.js';
import '../../ui/Tooltip.js';
import '../../ui/TooltipWrapper.js';
import '../../chunks/bundle-CBfyMFC9.js';
import '../../ui/MobileMenu.js';
import '../../chunks/bundle-DJrybJGG.js';
import '@sendbird/chat/message';
import '../../chunks/bundle-paSqms04.js';
import '../../chunks/bundle-ClIzLKH-.js';
import '../../chunks/bundle-DH45eb7x.js';
import '../../ui/OGMessageItemBody.js';
import '../../chunks/bundle-DJ_OSSMT.js';
import '../../ui/MentionLabel.js';
import '../../ui/LinkLabel.js';
import '../../ui/TextMessageItemBody.js';
import '../../ui/FileMessageItemBody.js';
import '../../ui/TextButton.js';
import '../../ui/VoiceMessageItemBody.js';
import '../../ui/ProgressBar.js';
import '../../VoicePlayer/useVoicePlayer.js';
import '../../chunks/bundle-DiCbzkpv.js';
import '../../VoiceRecorder/context.js';
import '../../ui/PlaybackTime.js';
import '../../ui/ThumbnailMessageItemBody.js';
import '../../chunks/bundle-BlbRAJmA.js';
import '../../ui/UnknownMessageItemBody.js';
import '../../ui/TemplateMessageItemBody.js';
import '../../chunks/bundle-MsVmXHII.js';
import '../../ui/FallbackTemplateMessageItemBody.tsx.js';
import '../../ui/LoadingTemplateMessageItemBody.tsx.js';
import '../../ui/MessageFeedbackFailedModal.js';
import '../../chunks/bundle-DDJKNqBp.js';
import '../../chunks/bundle-DVVKBU16.js';
import '../../Message/hooks/useDirtyGetMentions.js';
import '../../chunks/bundle-HzWL_k--.js';
import '../../chunks/bundle-CDYhXWNX.js';
import '../../chunks/bundle-vbkNCu0L.js';
import '../../chunks/bundle-CopK-w3W.js';

function ThreadList(_a) {
    var className = _a.className, _b = _a.renderMessage, renderMessage = _b === void 0 ? function (props) { return React__default.createElement(ThreadListItem, __assign({}, props)); } : _b, renderCustomSeparator = _a.renderCustomSeparator, scrollRef = _a.scrollRef, scrollBottom = _a.scrollBottom;
    var config = useSendbird().state.config;
    var stringSet = useLocalization().stringSet;
    var userId = config.userId;
    var _c = useThread().state, currentChannel = _c.currentChannel, allThreadMessages = _c.allThreadMessages, localThreadMessages = _c.localThreadMessages;
    return (React__default.createElement("div", { className: "sendbird-thread-list ".concat(className) },
        allThreadMessages.map(function (message, idx) {
            var _a;
            var isByMe = ((_a = message === null || message === void 0 ? void 0 : message.sender) === null || _a === void 0 ? void 0 : _a.userId) === userId;
            var prevMessage = allThreadMessages[idx - 1];
            var nextMessage = allThreadMessages[idx + 1];
            // eslint-disable-next-line no-constant-condition
            var _b = compareMessagesForGrouping(prevMessage, message, nextMessage, stringSet, currentChannel, getCaseResolvedReplyType(config.groupChannel.replyType).upperCase)
                , chainTop = _b[0], chainBottom = _b[1];
            var hasSeparator = !((prevMessage === null || prevMessage === void 0 ? void 0 : prevMessage.createdAt) > 0 && (isSameDay(message === null || message === void 0 ? void 0 : message.createdAt, prevMessage === null || prevMessage === void 0 ? void 0 : prevMessage.createdAt)));
            var handleScroll = function () {
                var current = scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current;
                if (current && scrollBottom) {
                    var bottom = current.scrollHeight - current.scrollTop - current.offsetHeight;
                    if (scrollBottom < bottom) {
                        current.scrollTop += bottom - scrollBottom;
                    }
                }
            };
            return (React__default.createElement(MessageProvider, { message: message, isByMe: isByMe, key: message === null || message === void 0 ? void 0 : message.messageId }, renderMessage({
                message: message,
                chainTop: chainTop,
                chainBottom: chainBottom,
                hasSeparator: hasSeparator,
                renderCustomSeparator: renderCustomSeparator,
                handleScroll: handleScroll,
            })));
        }),
        localThreadMessages.map(function (message) {
            var _a;
            var isByMe = ((_a = message === null || message === void 0 ? void 0 : message.sender) === null || _a === void 0 ? void 0 : _a.userId) === userId;
            var handleScroll = function () {
                var current = scrollRef === null || scrollRef === void 0 ? void 0 : scrollRef.current;
                if (current) {
                    var bottom = current.scrollHeight - current.scrollTop - current.offsetHeight;
                    if (scrollBottom < bottom) {
                        current.scrollTop += bottom - scrollBottom;
                    }
                }
            };
            return (React__default.createElement(MessageProvider, { message: message, isByMe: isByMe, key: message === null || message === void 0 ? void 0 : message.messageId }, renderMessage({
                message: message,
                hasSeparator: false,
                renderCustomSeparator: renderCustomSeparator,
                handleScroll: handleScroll,
            })));
        })));
}

export { ThreadList as default };
